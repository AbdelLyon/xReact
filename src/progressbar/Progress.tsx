import { useEffect, useState, useMemo } from "react";

import { Progress as ProgressBar } from "@/ui/progress";
import { cn } from "@/utils";

export type Segment = {
  value: number;
  color: string;
};

type ProgressProps = {
  segments: Segment[];
  animate?: boolean;
  animationDuration?: number;
  maxValue?: number; // New prop for maximum value
};

export function Progress({
  segments,
  animate = false,
  animationDuration = 500,
  maxValue, // Add maxValue to destructured props
}: Readonly<ProgressProps>) {
  const [progress, setProgress] = useState<number[]>([]);

  const totalValue = segments.reduce((acc, segment) => acc + segment.value, 0);
  const effectiveMaxValue = maxValue !== undefined ? maxValue : totalValue;

  const percentageSegments = useMemo(() => {
    return segments.map((segment) => ({
      ...segment,
      percentage: (segment.value / effectiveMaxValue) * 100,
    }));
  }, [segments, effectiveMaxValue]);

  useEffect(() => {
    if (animate) {
      setProgress(new Array(percentageSegments.length).fill(0));
      const timers = percentageSegments.map((segment, index) => {
        return setTimeout(() => {
          setProgress((prev) => {
            const newProgress = [...prev];

            newProgress[index] = segment.percentage;

            return newProgress;
          });
        });
      });

      return () => timers.forEach((timer) => clearTimeout(timer));
    } else {
      setProgress(percentageSegments.map((segment) => segment.percentage));
    }
  }, [effectiveMaxValue, segments, animate, percentageSegments]);

  return (
    <div className="w-full">
      <div className="relative h-4 w-full overflow-hidden rounded-full border border-border bg-gray-200">
        {percentageSegments.map((segment, index) => {
          const leftPosition = percentageSegments
            .slice(0, index)
            .reduce((sum, seg) => sum + seg.percentage, 0);

          return (
            <ProgressBar
              key={`progress-bar-${segment.color}-${index}`}
              className={cn("absolute h-full rounded-none ")}
              style={{
                left: `${leftPosition}%`,
                width: animate
                  ? `${progress[index]}%`
                  : `${segment.percentage}%`,
                backgroundColor: segment.color,
                transition: animate
                  ? `width ${animationDuration}ms ease-in`
                  : "none",
              }}
              value={progress[index]}
            />
          );
        })}
      </div>
    </div>
  );
}
