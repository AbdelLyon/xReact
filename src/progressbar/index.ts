import "@/index.css";
export { Progress, type Segment } from "@/progressbar/Progress";
