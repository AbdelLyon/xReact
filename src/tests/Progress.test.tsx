import { render, act } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";

import { Progress } from "@/progressbar";

vi.useFakeTimers();

describe("Progress", () => {
  it("renders correctly without animation", () => {
    const segments = [
      { value: 30, color: "red" },
      { value: 50, color: "blue" },
      { value: 20, color: "green" },
    ];

    const { container } = render(<Progress segments={segments} />);

    expect(container).toMatchSnapshot();
  });

  it("renders correctly with animation", () => {
    const segments = [
      { value: 30, color: "red" },
      { value: 50, color: "blue" },
      { value: 20, color: "green" },
    ];

    const { container } = render(
      <Progress animate={true} animationDuration={500} segments={segments} />,
    );

    act(() => {
      vi.runAllTimers();
    });

    expect(container).toMatchSnapshot();
  });
});
