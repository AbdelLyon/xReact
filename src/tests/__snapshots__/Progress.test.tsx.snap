// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Progress > renders correctly with animation 1`] = `
<div>
  <div
    class="w-full"
  >
    <div
      class="relative h-4 w-full overflow-hidden rounded-full border border-border bg-gray-200"
    >
      <div
        aria-valuemax="100"
        aria-valuemin="0"
        class="w-full overflow-hidden bg-secondary absolute h-full rounded-none"
        data-max="100"
        data-state="indeterminate"
        role="progressbar"
        style="left: 0%; background-color: red; transition: width 500ms ease-in; width: 30%;"
      >
        <div
          class="size-full flex-1 transition-all"
          data-max="100"
          data-state="indeterminate"
          style="left: 0%; background-color: red; transition: width 500ms ease-in; transform: translateX(-70%); width: 30%;"
        />
      </div>
      <div
        aria-valuemax="100"
        aria-valuemin="0"
        class="w-full overflow-hidden bg-secondary absolute h-full rounded-none"
        data-max="100"
        data-state="indeterminate"
        role="progressbar"
        style="left: 30%; background-color: blue; transition: width 500ms ease-in; width: 50%;"
      >
        <div
          class="size-full flex-1 transition-all"
          data-max="100"
          data-state="indeterminate"
          style="left: 30%; background-color: blue; transition: width 500ms ease-in; transform: translateX(-50%); width: 50%;"
        />
      </div>
      <div
        aria-valuemax="100"
        aria-valuemin="0"
        class="w-full overflow-hidden bg-secondary absolute h-full rounded-none"
        data-max="100"
        data-state="indeterminate"
        role="progressbar"
        style="left: 80%; background-color: green; transition: width 500ms ease-in; width: 20%;"
      >
        <div
          class="size-full flex-1 transition-all"
          data-max="100"
          data-state="indeterminate"
          style="left: 80%; background-color: green; transition: width 500ms ease-in; transform: translateX(-80%); width: 20%;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Progress > renders correctly without animation 1`] = `
<div>
  <div
    class="w-full"
  >
    <div
      class="relative h-4 w-full overflow-hidden rounded-full border border-border bg-gray-200"
    >
      <div
        aria-valuemax="100"
        aria-valuemin="0"
        class="w-full overflow-hidden bg-secondary absolute h-full rounded-none"
        data-max="100"
        data-state="indeterminate"
        role="progressbar"
        style="left: 0%; width: 30%; background-color: red; transition: none;"
      >
        <div
          class="size-full flex-1 transition-all"
          data-max="100"
          data-state="indeterminate"
          style="left: 0%; width: 30%; background-color: red; transition: none; transform: translateX(-70%);"
        />
      </div>
      <div
        aria-valuemax="100"
        aria-valuemin="0"
        class="w-full overflow-hidden bg-secondary absolute h-full rounded-none"
        data-max="100"
        data-state="indeterminate"
        role="progressbar"
        style="left: 30%; width: 50%; background-color: blue; transition: none;"
      >
        <div
          class="size-full flex-1 transition-all"
          data-max="100"
          data-state="indeterminate"
          style="left: 30%; width: 50%; background-color: blue; transition: none; transform: translateX(-50%);"
        />
      </div>
      <div
        aria-valuemax="100"
        aria-valuemin="0"
        class="w-full overflow-hidden bg-secondary absolute h-full rounded-none"
        data-max="100"
        data-state="indeterminate"
        role="progressbar"
        style="left: 80%; width: 20%; background-color: green; transition: none;"
      >
        <div
          class="size-full flex-1 transition-all"
          data-max="100"
          data-state="indeterminate"
          style="left: 80%; width: 20%; background-color: green; transition: none; transform: translateX(-80%);"
        />
      </div>
    </div>
  </div>
</div>
`;
